docs/006-just-in-time-channels.adoc

```diff
@@ -0,0 +1,59 @@
+= ADR 006 - Just-in-time channels
+Philipp Hoenisch <philipp@10101.finance>
+3.0, July 29, 2022: AsciiDoc article template
+:toc:
+:icons: font
+:attributes: 2023-02-13
+
+Status: Draft
+
+This document contains a description about the protocol we use for channel-less payments.
+
+== Decision
+
+Due to technical reasons in rust-lightning we opted for the following protocol:
+
+. Alice tells the coordinator that she is expecting a payment. The coordinator generates a `fake_channel_id`
+. Alice generates an invoice which includes a routing hint using this fake channel id
+. Alice gives the invoice to Bob
+. Bob pays the invoice
+. The coordinator intercepts this payment and checks for known `fake_channel_ids`
+. The coordinator opens a channel
+. Once the channel is open, the coordinator forwards the payment
+
+[source,mermaid]
+----
+sequenceDiagram
+
+title Payment Flow
+
+participant Alice
+participant Coordinator
+participant Bob
+
+Alice->>Coordinator: tells she is expecting a payment
+Coordinator->>Coordinator: generates `fake_channel_id`
+Coordinator->>Alice: fake_channel_id
+Alice->>Alice: generates invoice with routing hint using `fake_channel_id`
+Alice->>Bob: invoice
+Bob->>Bob: pays invoice
+Bob->>Coordinator: payment
+Coordinator->>Coordinator: intercepts payment
+Coordinator->>Coordinator: checks for known fake_channel_id
+Coordinator->>Coordinator: opens channel
+Coordinator->>Alice: payment
+----
+
+== Context
+
+We want to offer to the users of the 10101-app a great UX when initially transferring funds into the wallet.
+For example, Alice uses 10101 and wants to receive a payment from Bob via the coordinator but does not yet have a channel open with the coordinator.
+To receive the payment, the coordinator will have to intercept the payment, create a channel on-the-fly to Alice and then forward the payment.
+
+The same feature slightly adapted can be used later-on when receiving funds which exceed the in-bound channel capacity.
+
+=== Options considered
+
+We initially thought that for the described feature above, no communication is needed between the coordinator and the payee (Alice in the scenario above).
+However, due to technical constraints in https://github.com/lightningdevkit/rust-lightning/blob/a1f7cbb572fa50dcb933bb162facb5ce2cd82576/lightning/src/util/events.rs#L753-L784[rust-lightning], we do not learn about the next hop ID.
```

Comment:
I wonder if there is a way to tie the fake SCID to the payee public key without the communication step. After all, the coordinator knows the public keys of all the peers that they're connected to. Maybe we could come up with a deterministic algorithm to derive the fake SCID from the public key?

In any case, this is for sure good enough at this stage, and perhaps forever.

Reply:
Potentially, but we would need to dig into rust-lightning for that.

Reply:
I thought that wouldn't be necessary!

1. Alice generates fake SCID with a deterministic algorithm based on her PK.
2. The coordinator eventually learns about it.
3. The coordinator applies the same algorithm to the PKs of all the connected peers without a channel.
4. Only one result matches the fake SCID from the event.
5. The coordinator then attempts to open the channel with the identified peer.

In my mind, steps (3), (4) and (5) happen in `ln_dlc_node`. Obviously, step (4) has to work reliably.

Reply:
ah, I think now I get it. Sorry, I was confused. 

I think you are right, this should be possible. 

Note that the currently used `get_intercept_scid` is doing quite some stuff to generate a _realistic looking_ scid.

https://github.com/lightningdevkit/rust-lightning/blob/ac6e0b3fed4119c196a5e1e864239087422727ea/lightning/src/util/scid_utils.rs#L100-L104

Reply:
Wow, okay. Good point. Let's not mess with this for now. It might not even be a real problem anyway.